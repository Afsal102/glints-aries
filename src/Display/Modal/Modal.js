/* @flow */

import React, { Component, Fragment } from 'react';

import Icon from '../../General/Icon';

import { escEvent } from '../../Utils/DomUtils';

import {
  ModalContainer,
  ModalDialog,
  ModalContentArea,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from '../../Style/Display/ModalStyle';

class Modal extends Component <Props, State> {
  state = {
    isOpen: false,
  }

  componentDidMount() {
    const { onClose } = this.props;
    document.addEventListener('keydown', escEvent(onClose), false);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.isVisible && !prevState.isOpen) {
      document.body.style.overflow = 'hidden';
      return { isOpen: true };
    }

    if (!nextProps.isVisible && prevState.isOpen) {
      document.body.removeAttribute('style');
      return { isOpen: false };
    }
    return null;
  }

  componentDidUpdate(prevProps) {
    const { isVisible } = this.props;
    if (!prevProps.isVisible && isVisible) {
      document.getElementById('modal-content').focus();
    }
  }

  componentWillUnmount() {
    const { onClose } = this.props;
    document.removeEventListener('keydown', escEvent(onClose), false);
    document.body.removeAttribute('style');
  }

  render() {
    const {
      title,
      onClose,
      type,
      children,
      className,
      hideContentArea,
      centering,
      removeAnimation,
      footer,
      size,
      hideHeader,
      ...defaultProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <ModalContainer
        id="aries-modal"
        className={className}
        centering={centering}
        onClick={() => onClose()}
        isOpen={isOpen}
        removeAnimation={removeAnimation}
      >
        <ModalDialog id="modal-dialog">
          <ModalContentArea
            id="modal-content"
            role="dialog"
            aria-modal="true"
            hideContentArea={hideContentArea}
            centering={centering}
            onClick={e => e.stopPropagation()}
            tabIndex={0}
            isOpen={isOpen}
            removeAnimation={removeAnimation}
            size={size}
            {...defaultProps}
          >
            <If condition={!hideHeader}>
              <ModalHeader id="modal-header">
                <h3>{ title }</h3>
                <button type="button" onClick={() => onClose()}>
                  <Icon name="close" color={hideContentArea ? 'white' : 'grey'} />
                </button>
              </ModalHeader>
            </If>
            <ModalBody
              id="modal-body"
              hideContentArea={hideContentArea}
            >
              {children}
            </ModalBody>
            <If condition={footer !== undefined}>
              <ModalFooter id="modal-footer">
                { footer.map(data => React.cloneElement(data, { key: data.props.children })) }
              </ModalFooter>
            </If>
          </ModalContentArea>
        </ModalDialog>
      </ModalContainer>
    );
  }
}

type Props = {
  children: React$Node,
  title: string,
  isVisible: boolean,
  type: string,
  onClose: Function,
  className: string,
  hideContentArea: boolean,
  centering: boolean,
  removeAnimation: boolean,
  footer: React$Node,
  size: string,
  hideHeader: boolean,
}

type State = {
  isOpen: boolean,
}

export default Modal;
