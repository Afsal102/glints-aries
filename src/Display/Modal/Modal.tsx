import * as React from 'react';
import classNames from 'classnames';

import { CloseIcon } from '../../General/Icon/components';
import { escEvent } from '../../Utils/DomUtils';

import {
  ModalContainer,
  ModalDialog,
  ModalContentArea,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from '../../Style/Display/ModalStyle';

class Modal extends React.Component<Props, State> {
  modalContentAreaRef: React.RefObject<HTMLDivElement>;
  mouseDownTarget: HTMLElement;
  escEvent: (e: KeyboardEvent) => any;

  state = {
    isOpen: false,
  };

  constructor(props: Props) {
    super(props);
    this.modalContentAreaRef = React.createRef();
  }

  guardedOnClose() {
    const { onClose } = this.props;
    if (this.state.isOpen) onClose();
  }

  componentDidMount() {
    this.escEvent = escEvent(this.guardedOnClose.bind(this));
    document.addEventListener('keydown', this.escEvent, false);
  }

  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
    if (nextProps.isVisible && !prevState.isOpen) {
      document.body.style.overflow = 'hidden';
      return { isOpen: true };
    }

    if (!nextProps.isVisible && prevState.isOpen) {
      document.body.removeAttribute('style');
      return { isOpen: false };
    }
    return null;
  }

  componentDidUpdate(prevProps: Props) {
    const { isVisible } = this.props;
    if (!prevProps.isVisible && isVisible) {
      this.modalContentAreaRef.current.focus();
    }
  }

  removeEscapeEventListener() {
    document.removeEventListener('keydown', this.escEvent, false);
  }

  componentWillUnmount() {
    this.removeEscapeEventListener();
    document.body.removeAttribute('style');
  }

  // To prevent the modal from closing
  // when a mousedown event occurs inside the ModalContentArea
  // but the subsequent mouseup event occurs outside
  handleMouseDown = (e: React.MouseEvent) => {
    const element = e.target as HTMLElement;
    this.mouseDownTarget = element;
  };

  handleClick = (e: React.MouseEvent) => {
    const element = e.target as HTMLElement;
    if (this.mouseDownTarget === element) {
      this.props.onClose();
    }
  };

  render() {
    const {
      title,
      onClose,
      children,
      className,
      hideContentArea,
      centering,
      removeAnimation,
      footer,
      size,
      hideHeader,
      ...restProps
    } = this.props;

    const { isOpen } = this.state;

    return (
      <ModalContainer
        data-testid="modal-container"
        className={classNames('aries-modal', className)}
        centering={centering}
        isOpen={isOpen}
        removeAnimation={removeAnimation}
        onMouseDown={this.handleMouseDown}
        onClick={this.handleClick}
      >
        <ModalDialog className="modal-dialog">
          <ModalContentArea
            className="modal-content"
            role="dialog"
            aria-modal="true"
            hideContentArea={hideContentArea}
            centering={centering}
            onClick={e => e.stopPropagation()}
            tabIndex={0}
            isOpen={isOpen}
            removeAnimation={removeAnimation}
            size={size}
            ref={this.modalContentAreaRef}
            {...restProps}
          >
            {!hideHeader && (
              <ModalHeader className="modal-header">
                <h3>{title}</h3>
                <button
                  data-testid="close-button"
                  type="button"
                  onClick={onClose}
                >
                  <CloseIcon color={hideContentArea ? 'white' : 'grey'} />
                </button>
              </ModalHeader>
            )}
            <ModalBody className="modal-body" hideContentArea={hideContentArea}>
              {children}
            </ModalBody>
            {footer !== undefined && (
              <ModalFooter className="modal-footer">{footer}</ModalFooter>
            )}
          </ModalContentArea>
        </ModalDialog>
      </ModalContainer>
    );
  }
}

export type sizeType = 's' | 'm' | 'l' | 'xl';

interface Props
  extends React.ComponentPropsWithoutRef<typeof ModalContentArea> {
  children: React.ReactNode;
  title?: string;
  isVisible: boolean;
  onClose(): void;
  hideContentArea?: boolean;
  centering?: boolean;
  removeAnimation?: boolean;
  footer?: React.ReactElement[];
  size?: sizeType;
  hideHeader?: boolean;
}

interface State {
  isOpen: boolean;
}

export default Modal;
