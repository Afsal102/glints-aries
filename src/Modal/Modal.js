/* @flow */

import React, { Component, Fragment } from 'react';
import { Container, ModalBackground, ModalArea } from '../Style/ModalStyle';

class Modal extends Component <Props> {
  render() {
    const {
      visibility,
      close,
      type,
      children,
      className,
      hideContentArea,
      ...defaultProps
    } = this.props;

    return (
      <Fragment>
        {visibility && (
          <Container {...defaultProps} className={className}>
            <ModalBackground onClick={close} className="modalBackground" />
            <ModalArea className="modalArea" type={type} hideContentArea={hideContentArea}>
              {children}
            </ModalArea>
          </Container>
        )}
      </Fragment>
    );
  }
}

type Props = {
  children: React$Node,
  visibility: boolean,
  type: string,
  close: Function,
  className: string,
  hideContentArea: boolean,
}

export default Modal;
