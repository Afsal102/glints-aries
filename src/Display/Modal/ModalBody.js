/* @flow */

import React from 'react';
import { ModalBodyWrapper } from '../../Style/Display/ModalStyle';

const ModalBody = (props: Props) => {
  const {
    children,
    className,
    ...defaultProps
  } = props;

  return (
    <ModalBodyWrapper className={className} {...defaultProps}>
      { children }
    </ModalBodyWrapper>
  );
};

type Props = {
  children: React$Node,
  className: string,
}

export default ModalBody;
