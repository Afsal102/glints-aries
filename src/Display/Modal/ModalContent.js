/* @flow */

import React from 'react';
import { ModalContentStyle } from '../../Style/Display/ModalStyle';

const ModalContent = (props: Props) => {
  const {
    children,
    className,
    ...defaultProps
  } = props;

  return (
    <ModalContentStyle className={className} {...defaultProps}>
      { children }
    </ModalContentStyle>
  );
};

type Props = {
  children: React$Node,
  className: string,
}

export default ModalContent;
