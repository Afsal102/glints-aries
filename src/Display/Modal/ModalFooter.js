/* @flow */

import React from 'react';
import { ModalFooterStyle } from '../../Style/Display/ModalStyle';

const ModalFooter = (props: Props) => {
  const {
    children,
    className,
    ...defaultProps
  } = props;

  return (
    <ModalFooterStyle className={className} {...defaultProps}>
      { children }
    </ModalFooterStyle>
  );
};

type Props = {
  children: React$Node,
  className: string,
}

export default ModalFooter;
