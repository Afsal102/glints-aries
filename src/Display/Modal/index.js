// @flow

import Modal from './Modal';
import ModalBody from './ModalBody';
import ModalContent from './ModalContent';
import ModalFooter from './ModalFooter';

Modal.Body = ModalBody;
Modal.Content = ModalContent;
Modal.Footer = ModalFooter;

export default Modal;
