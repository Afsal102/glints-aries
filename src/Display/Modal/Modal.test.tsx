import * as React from 'react';
import Modal from '.';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

const props = {
  title: 'Modal Title',
  content: 'Modal Content',
  footer: 'Modal Footer',
  onClose: jest.fn(),
};

const OpenedModal = (
  <Modal
    title={props.title}
    isVisible={true}
    footer={[<button key="button">{props.footer}</button>]}
    onClose={props.onClose}
  >
    <p>{props.content}</p>
  </Modal>
);

function setupModal(isVisible: boolean) {
  const onClose = jest.fn();
  const ModalComponent = (
    <Modal
      title={props.title}
      isVisible={isVisible}
      footer={[<button key="button">{props.footer}</button>]}
      onClose={onClose}
    >
      <p>{props.content}</p>
    </Modal>
  );
  const { getByTestId, getByRole } = render(ModalComponent);
  const modalContainer = getByTestId('modal-container');
  const modalDialog = getByRole('dialog');
  return [modalContainer, onClose, modalDialog];
}

it('<Modal> should render with a title, content, footer and an onClick handler', () => {
  const ModalSnapshot = renderer.create(OpenedModal).toJSON();
  expect(ModalSnapshot).toMatchSnapshot();
});

describe('when modal is opened', () => {
  it('should be visible', () => {
    const [modalContainer] = setupModal(true);
    expect(modalContainer).toHaveStyle('visibility: visible');
  });

  it('should not be closed if the modal is clicked', () => {
    const [, onClose, modalDialog] = setupModal(true);
    fireEvent.click(modalDialog as Element);
    expect(onClose).toHaveBeenCalledTimes(0);
  });
});

describe('when modal is opened', () => {
  it('should display the title', () => {
    const { getByText } = render(OpenedModal);
    const modalHeader = getByText(props.title);
    expect(modalHeader).toHaveTextContent(props.title);
  });

  it('should display the content', () => {
    const { getByText } = render(OpenedModal);
    const modalContent = getByText(props.content);
    expect(modalContent).toHaveTextContent(props.content);
  });

  it('should display the footer', () => {
    const { getByText } = render(OpenedModal);
    const modalFooter = getByText(props.footer);
    expect(modalFooter).toHaveTextContent(props.footer);
  });
});

describe('when modal is closed', () => {
  it('should not be visible', () => {
    const [modalContainer] = setupModal(false);
    expect(modalContainer).toHaveStyle('visibility: hidden');
  });
});

describe('onClose should be called once when:', () => {
  it('area outside modal is clicked', () => {
    const [modalContainer, onClose] = setupModal(true);
    fireEvent.click(modalContainer as Element);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('escape key is pressed', () => {
    const [modalContainer, onClose] = setupModal(true);
    fireEvent.keyDown(modalContainer as Element, {
      key: 'Escape',
      keyCode: 27,
    });
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});
