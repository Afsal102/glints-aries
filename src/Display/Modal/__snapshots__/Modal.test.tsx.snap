// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal> should render with a title, content, footer and an onClick handler 1`] = `
<div
  className="aries-modal sc-bdVaJa aauuo"
  data-testid="modal-container"
  onClick={[Function]}
>
  <div
    className="modal-dialog sc-bwzfXH ghuPdm"
  >
    <div
      aria-modal="true"
      className="modal-content sc-htpNat lfPwe"
      onClick={[Function]}
      role="dialog"
      tabIndex={0}
    >
      <header
        className="modal-header sc-bxivhb yXCPR"
      >
        <h3>
          Modal Title
        </h3>
        <button
          data-testid="close-button"
          onClick={[Function]}
          type="button"
        >
          <svg
            fill="grey"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 10L90 0 50 40 10 0 0 10l40 40L0 90l10 10 40-40 40 40 10-10-40-40z"
            />
          </svg>
        </button>
      </header>
      <section
        className="modal-body sc-ifAKCX cwNdkR"
      >
        <p>
          Modal Content
        </p>
      </section>
      <footer
        className="modal-footer sc-EHOje tLqWc"
      >
        <button>
          Modal Footer
        </button>
      </footer>
    </div>
  </div>
</div>
`;
